libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs

```diff
@@ -0,0 +1,217 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Cognitive.LUIS;
+using Microsoft.Cognitive.LUIS.Models;
+using Newtonsoft.Json.Linq;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Bot.Builder.LUIS
+{
+    /// <inheritdoc />
+    /// <summary>
+    /// A Luis based implementation of IRecognizer
+    /// </summary>
+    public class LuisRecognizer : IRecognizer
+    {
+        private readonly LuisService _luisService;
+        private readonly ILuisOptions _luisOptions;
+        private readonly ILuisRecognizerOptions _luisRecognizerOptions;
+
+        public LuisRecognizer(ILuisModel luisModel, ILuisRecognizerOptions luisRecognizerOptions = null, ILuisOptions options = null)
+        {
+            _luisService = new LuisService(luisModel);
+            _luisOptions = options;
+            _luisRecognizerOptions = luisRecognizerOptions;
+        }
+
+        /// <inheritdoc />
+        public Task<RecognizerResult> Recognize(string utterance, CancellationToken ct)
+        {
+            var luisRequest = new LuisRequest(utterance);
+            _luisOptions?.Apply(luisRequest);
+            return Recognize(luisRequest, ct, _luisRecognizerOptions?.Verbose ?? true);
+        }
+
+        private async Task<RecognizerResult> Recognize(LuisRequest request, CancellationToken ct, bool verbose)
+        {
+            var luisResult = await _luisService.QueryAsync(request, ct);
+
+            var recognizerResult = new RecognizerResult
+            {
+                Text = request.Query,
+                Intents = GetIntents(luisResult),
+                Entities = GetEntitiesAndMetadata(luisResult.Entities, luisResult.CompositeEntities, verbose)
+            };
+
+            return recognizerResult;
+        }
+
+        private static JObject GetIntents(LuisResult luisResult)
```

Comment:
Consider simplifying this code, so it's not doing so many things on a single line.

Reply:
I think it's easy to follow as-is, and VS now supports breakpoints inside line so debugging it is easy too. I find this functional style easier to follow them imperative code and looping when there isn't really business logic but just transformations of data.